google/cloud/bigquery/dbapi/cursor.py

```diff
@@ -280,6 +281,8 @@ def _bqstorage_fetch(self, bqstorage_client):
         # bigquery_storage can indeed be imported here without errors.
         from google.cloud import bigquery_storage
 
+        _verify_bq_storage_version()
```

Comment:
If we do the check in `Connection.__init__`, why do we need to do it here?  Are we afraid that someone would set that attribute?

Why is it fatal here but not in `Connection.__init__`?

Reply:
Somebody could create a `Cursor` instance directly, i.e. without going through `Connection.cursor()`:

```py
conn = dbapi.Connection()
cursor = dbapi.Cursor(conn)
```

Not saying this is necessarily a recommended pattern, but `Cursor` is directly exposed in `google.cloud.bigquery.dbapi` after all, thus has its own check.

Reply:
> Somebody could create a `Cursor` instance directly, i.e. without going through `Connection.cursor()`:
> 
> ```python
> conn = dbapi.Connection()
> cursor = dbapi.Cursor(conn)
> ```

Sure, but the `Connection` constructor will still do the check.  The cursor will still get the storage client from the connection.

I may be misunderstanding you. :)

Reply:
Or maybe it's too late here already. :)
Yeah, you're right, `Cursor` takes the client that's stored on the `Connection` instance, will simplify. And I even think there's still one non-graceful code path in `dbpai` (read; no fallback on the REST client), will double check tomorrow.

Reply:
> > Why is it fatal here but not in `Connection.__init__`?
> 
> This method specifically tries to fetch data over the BQ Storage API, and if it is given an outdated client, it's the caller's fault somewhere higher up the call chain. The method can't guess how to recover, thus at least throws a more informative exception.
> 
> (otherwise an `AttributeError` would be raised a few lines below at `bigquery_storage.types.DataFormat.ARROW`)

See my straw man I just posted for `_create_bqstorage_client`.  If all `bqstorage_client` assignments go through that, then you can never end up with an invalid client and thus never get the `AttributeError` mentioned above.

From your other comments, I think you can see where I'm going. :)